src/constraints.jl

```diff
@@ -0,0 +1,559 @@
+                       (:DisjunctionRef, :disjunctions), 
+                       (:LogicalConstraintRef, :logical_constraints))
+    @eval begin
+        @doc """
+            JuMP.owner_model(cref::$($RefType))
+
+        Return the model to which `cref` belongs.
+        """
+        JuMP.owner_model(cref::$RefType) = cref.model
+
+        @doc """
+            JuMP.index(cref::$($RefType))
+
+        Return the index constraint associated with `cref`.
+        """
+        JuMP.index(cref::$RefType) = cref.index
+
+        @doc """
+            JuMP.is_valid(model::JuMP.Model, cref::$($RefType))
+
+        Return `true` if `cref` refers to a valid constraint in the `GDP model`.
+        """
+        function JuMP.is_valid(model::JuMP.Model, cref::$RefType) # TODO: generalize for AbstractModel
+            return model === JuMP.owner_model(cref)
+        end
+
+        # Get the ConstraintData object
+        function _constraint_data(cref::$RefType)
+            return gdp_data(JuMP.owner_model(cref)).$loc[JuMP.index(cref)]
+        end
+
+        @doc """
+            JuMP.name(cref::$($RefType))
+
+        Get a constraint's name attribute.
+        """
+        function JuMP.name(cref::$RefType)
+            return _constraint_data(cref).name
+        end
+
+        @doc """
+            JuMP.set_name(cref::$($RefType), name::String)
+
+        Set a constraint's name attribute.
+        """
+        function JuMP.set_name(cref::$RefType, name::String)
+            _constraint_data(cref).name = name
+            _set_ready_to_optimize(JuMP.owner_model(cref), false)
+            return
+        end
+
+        @doc """
+            JuMP.constraint_object(cref::$($RefType))
+
+        Return the underlying constraint data for the constraint 
+            referenced by `cref`.
+        """
+        function JuMP.constraint_object(cref::$RefType)
+            return _constraint_data(cref).constraint
+        end
+
+        # Extend Base methods
+        function Base.:(==)(cref1::$RefType, cref2::$RefType)
+            return cref1.model === cref2.model && cref1.index == cref2.index
+        end
+        Base.copy(cref::$RefType) = cref
+        @doc """
+            Base.getindex(map::JuMP.GenericReferenceMap, cref::$($RefType))
+
+        ...
+        """
+        function Base.getindex(map::JuMP.ReferenceMap, cref::$RefType)
+            $RefType(map.model, JuMP.index(cref))
+        end
+    end
+end
+
+# Extend delete
+"""
+    JuMP.delete(model::JuMP.Model, cref::DisjunctionRef)
+
+Delete a disjunction constraint from the `GDP model`.
+"""
+function JuMP.delete(model::JuMP.Model, cref::DisjunctionRef)
+    @assert JuMP.is_valid(model, cref) "Disjunctive constraint does not belong to model."
+    cidx = JuMP.index(cref)
+    dict = _disjunctions(model)
+    #delete each logical variable
+    for lv_idx in dict[cidx].constraint.disjuncts
+        JuMP.delete(model, LogicalVariableRef(model, lv_idx))
+    end
```

Comment:
Since we are forcing users to create logical variables and associate them with disjunction constraints, wouldn't deleting them here invalidate the constraints that aren't deleted?

Reply:
Ok that  makes sense